Multi ledger

matthiasgeihs opened the pull request:
Enables multi-ledger channels.

The following changes are applied:
- Implement special Asset type that contains `LedgerID`.
- Implement multi funder (adjudicator) that contain a set of funders (adjudicators) which are called depending on which assets a channel has.
- Revise watcher: The watcher ensures that in case of a multi-ledger channel, all watched ledgers are kept synchronized.

RmbRT commented on watcher/local/watcher.go:
```diff
@@ -354,14 +364,61 @@ func (ch *ch) handleEventsFromChain(registerer channel.Registerer, progresser ch
+	// Even if older states have been registered for more than one
+	// channel in the same family, the channel tree will be registered
+	// only once.
+	parent.subChsAccess.Lock()
```
does this lock recursively for subchannels? I think it's strange to sometimes lock this channel's sub-channels, and sometimes only lock this channel's parent's sub-channels. I can't validate this without some additional effort.

RmbRT commented on watcher/local/watcher.go:
```diff
@@ -354,14 +364,61 @@ func (ch *ch) handleEventsFromChain(registerer channel.Registerer, progresser ch
+	latestTx := ch.txRetriever.retrieve()
+	log.Debugf("Latest version is (%d)", latestTx.Version)
+
+	higherVersionAvailable := e.Version() < latestTx.Version && e.Version() >= ch.registeredVersion
+	unregisteredMultiLedger := ch.multiLedger && (!ch.registered || ch.registeredVersion < e.Version())
```
I would appreciate if the reason why the latestTx version is only used in the first check but not in the second check was documented in a comment.

matthiasgeihs replied:
Added a comment.

RmbRT commented on watcher/local/watcher.go:
```diff
@@ -354,14 +364,61 @@ func (ch *ch) handleEventsFromChain(registerer channel.Registerer, progresser ch
+			return
+		}
+		log.Debug("Registered successfully")
+		ch.registered = true
```
Why is ch.registered only set here?

matthiasgeihs replied:
This is the only place where we call `registerDispute`. Where would you expect to set registered elsewhere?

RmbRT commented:
As I addressed above, I think the general architecture and protocols need to be adapted a little more to actually support multi-ledger channels. Architecture-wise, it looks good to me. As I also suggested in a meeting, it would maybe be expedient if we migrated the core to multi-ledger but added a legacy single-ledger adapter for older backends.
Also, even when only rolling out non-force-move channels, the on-chain contracts would need an option in the channel params that allows the user to disable force-moving.

matthiasgeihs commented on backend/multi/asset.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+// LedgerIDs returns the identifiers of the associated ledgers.
+func (a assets) LedgerIDs() ([]LedgerID, error) {
```
This is a helper function. No specific order is expected. The return value should be something that can be iterated on so I decided to use a slice because it seemed clean to me. If you insist, I can also return the map instead.

matthiasgeihs commented on backend/multi/funder.go:
```diff
@@ -0,0 +1,91 @@
+	errs := make(chan error, n)
+	for _, l := range ledgers {
+		go func(l LedgerID) {
+			err := func() error {
```
ok

matthiasgeihs commented on client/test/backend.go:
```diff
@@ -47,22 +49,39 @@ type (
 	return &MockBackend{
 		log:          log.Default(),
-		rng:          newThreadSafePrng(rng),
+		rng:          newThreadSafePrng(backendRng),
```
I can't control the spawning of go-routines. They are spawned outside of the mock backend.

matthiasgeihs commented on watcher/local/watcher.go:
```diff
@@ -295,50 +330,25 @@ func readPendingTxs(statesSub statesSub, timeout time.Duration) (channel.Transac
+			ch.handleRegisteredEvent(ctx, e, registerer, chRegistry)
+		case *channel.ProgressedEvent:
+			log.Debugf("Received progressed event from chain: %v", e)
+			if ch.multiLedger && e.Version() < ch.registeredVersion {
```
What I am addressing here is the case that a force move is only applied on chain A but not on chain B. In this case the watcher would make sure that the chains go back in sync. However, as you noted in our meeting, this does not address all attack vectors of multi-chain forced execution. The problem is that a malicious participant could force chain A into a different state than chain B with the same version number. Such a difference would have to be resolved with some new protocol, I believe.

As discussed, the easy solution for now is to disable forced execution for multi-ledger app channels.

matthiasgeihs commented on watcher/local/watcher.go:
```diff
@@ -354,14 +364,61 @@ func (ch *ch) handleEventsFromChain(registerer channel.Registerer, chRegistry *r
+	latestTx := ch.txRetriever.retrieve()
+	log.Debugf("Latest version is (%d)", latestTx.Version)
+
+	higherVersionAvailable := e.Version() < latestTx.Version && e.Version() >= ch.registeredVersion
+	unregisteredMultiLedger := ch.multiLedger && (!ch.registered || ch.registeredVersion < e.Version())
```
done

matthiasgeihs commented on backend/multi/dispatch.go:
```diff
@@ -0,0 +1,50 @@
+	"perun.network/go-perun/channel"
+)
+
+func (a *Adjudicator) dispatch(ledgers []LedgerID, f func(channel.Adjudicator) error) error {
```
no -> moved

matthiasgeihs commented on backend/multi/dispatch.go:
```diff
@@ -0,0 +1,50 @@
+	"perun.network/go-perun/channel"
+)
+
+func (a *Adjudicator) dispatch(ledgers []LedgerID, f func(channel.Adjudicator) error) error {
```
added a comment, the name `dispatch` is fine, IMO.

matthiasgeihs commented on channel/adjudicator.go:
```diff
@@ -25,7 +25,7 @@ import (
 
 //go:generate mockery --name AdjudicatorSubscription --output ../watcher/internal/mocks
-//go:generate mockery --name RegisterSubscriber --output ../watcher/internal/mocks
+//go:generate mockery --name WatcherAdjudicator --output ../watcher/internal/mocks
```
@manoranjith introduced them for testing the local watcher. I need to adopt the interface for the watcher, if we were implementing watching of multi-ledger app channels. However, after @RmbRT comments I decided to focus on non-app channels first and reverted many of the mock changes. However, the mocks were still generated incorrectly, because the package and directory needs to be named `mocks` (instead of `mock`).

matthiasgeihs commented on channel/adjudicator.go:
```diff
@@ -244,7 +242,7 @@ func NewRegisteredEvent(id ID, timeout Timeout, version uint64, state *State, si
 
 // NewProgressedEvent creates a new ProgressedEvent.
-func NewProgressedEvent(id ID, timeout Timeout, state *State, idx Index) *ProgressedEvent {
+func NewProgressedEvent(id ID, timeout Timeout, state *State, sig wallet.Sig, idx Index) *ProgressedEvent {
```
Not needed anymore as we focus on multi-ledger no app channels first.

matthiasgeihs commented on client/multiledger_happy_test.go:
```diff
@@ -0,0 +1,262 @@
+	"math/rand"
+	"testing"
+
+	"github.com/pkg/errors"
```
We don't strictly apply any rule for our imports currently. If you want, you can install a linter for that that checks that.

cryptphil replied:
The `goimports` linter should actually be active with `golangci-linter`, I wonder why it is not reporting this...

matthiasgeihs commented on client/multiledger_happy_test.go:
```diff
@@ -0,0 +1,262 @@
+	// Setup proposal handler.
+	channels := make(chan *client.Channel, 1)
+	errs := make(chan error)
+	var channelHandler client.ProposalHandlerFunc = func(cp client.ChannelProposal, pr *client.ProposalResponder) {
```
Currently this function accesses variables from the surrounding scope. I think it's OK if we keep it like that for the sake of simplicity.

matthiasgeihs commented on watcher/local/watcher.go:
```diff
@@ -195,27 +206,51 @@ func (w *Watcher) startWatching(
+	hasMulti := false
+	var id multi.LedgerID
+	for _, asset := range a {
+		multiAsset, ok := asset.(multi.Asset)
+		switch {
+		case !ok:
+			continue
+		case !hasMulti:
+			hasMulti = true
+			id = multiAsset.LedgerID()
+		case id.MapKey() != multiAsset.LedgerID().MapKey():
+			return true
+		}
```
The linter wanted me to change it into a switch which apparently is idiomatic.

matthiasgeihs commented on watcher/local/watcher.go:
```diff
@@ -183,7 +193,8 @@ func (w *Watcher) startWatching(
 		statesPubSub = newStatesPubSub()
 		eventsToClientPubSub = newAdjudicatorEventsPubSub()
-		return newCh(id, parent, signedState.Params, eventsFromChainSub, eventsToClientPubSub, statesPubSub), nil
+		multiLedger := assets(signedState.State.Assets).MultiLedger()
```
ok

RmbRT replied:
I think it's alright to name bools `x` instead of `hasX` or `isX`. But in this case, it looks as if you're querying a `MultiLedger` object.

matthiasgeihs replied:
OK, revised the naming and merged it with the functionality used in the proposal.

matthiasgeihs commented:
I have addressed all comments. Significant changes:

- Restrict usage of multi-ledger to non-app channels.
- Moved package `multi` from `backend` to `channel`.

cryptphil commented on client/multiledger_dispute_test.go:
```diff
@@ -0,0 +1,122 @@
+	// Store state.
+	req1 := client.NewTestChannel(chAliceBob).AdjudicatorReq()
+
+	// Alice registers state on l1 adjudicator.
```
More a general question about this test:
Would more branches be tested if there would be 2 versions and Alice first registers the older one and then Bob registers the newer one?
If this is the case, maybe it is reasonable to extend the dispute test.

RmbRT replied:
As I understand it, this is only for testing the multi-ledger client-side functionality, as the ledgers are expected to be individually fully tested already. I think it's good to have such a simplistic test.

cryptphil replied:
I see.

matthiasgeihs commented on channel/multi/funder.go:
```diff
@@ -0,0 +1,90 @@
+func (f *Funder) Fund(ctx context.Context, request channel.FundingReq) error {
+	// We wait for the funding timeout in a go routine and cancel the funding
+	// context if the timeout elapses.
+	ctx, cancel := context.WithCancel(ctx)
```
Good point. Implemented.

matthiasgeihs commented on channel/multi/subscription.go:
```diff
@@ -0,0 +1,80 @@
+	return <-s.events
+}
+
+// Err returns an error if one occurred.
```
Adapted the comment.

matthiasgeihs commented:
Addressed all comments. Here is the diff: https://github.com/hyperledger-labs/go-perun/compare/5c0ec670f3b19ab88de2a54925f8fd6575374c42..aeae7717a81d16cd268add2d769f6176814e49c6

Please check and approve if you are happy. @cryptphil @RmbRT
